components/brave_wallet/browser/internal/hd_key.cc

```diff
@@ -750,38 +713,43 @@ void HDKey::GeneratePublicKey() {
   }
 }
 
-std::string HDKey::Serialize(
-    uint32_t version,
-    std::unique_ptr<std::vector<uint8_t>, SecureZeroVectorDeleter<uint8_t>> key)
-    const {
+// https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#serialization-format
+std::string HDKey::Serialize(uint32_t version,
+                             base::span<const uint8_t> key) const {
   // version(4) || depth(1) || parent_fingerprint(4) || index(4) || chain(32) ||
-  // key(33)
-  std::unique_ptr<std::vector<uint8_t>, SecureZeroVectorDeleter<uint8_t>> buf(
-      new std::vector<uint8_t>, SecureZeroVectorDeleter<uint8_t>());
+  // key(32 or 33)
+  SecureVector buf;
+
+  buf.reserve(SERIALIZATION_LEN);
 
-  buf->push_back((version >> 24) & 0xFF);
-  buf->push_back((version >> 16) & 0xFF);
-  buf->push_back((version >> 8) & 0xFF);
-  buf->push_back((version >> 0) & 0xFF);
+  buf.push_back((version >> 24) & 0xFF);
+  buf.push_back((version >> 16) & 0xFF);
+  buf.push_back((version >> 8) & 0xFF);
+  buf.push_back((version >> 0) & 0xFF);
 
-  buf->push_back(depth_);
+  buf.push_back(depth_);
 
-  buf->push_back((parent_fingerprint_ >> 24) & 0xFF);
-  buf->push_back((parent_fingerprint_ >> 16) & 0xFF);
-  buf->push_back((parent_fingerprint_ >> 8) & 0xFF);
-  buf->push_back(parent_fingerprint_ & 0xFF);
+  buf.push_back((parent_fingerprint_ >> 24) & 0xFF);
+  buf.push_back((parent_fingerprint_ >> 16) & 0xFF);
+  buf.push_back((parent_fingerprint_ >> 8) & 0xFF);
+  buf.push_back(parent_fingerprint_ & 0xFF);
 
-  buf->push_back((index_ >> 24) & 0xFF);
-  buf->push_back((index_ >> 16) & 0xFF);
-  buf->push_back((index_ >> 8) & 0xFF);
-  buf->push_back(index_ & 0xFF);
+  buf.push_back((index_ >> 24) & 0xFF);
+  buf.push_back((index_ >> 16) & 0xFF);
+  buf.push_back((index_ >> 8) & 0xFF);
+  buf.push_back(index_ & 0xFF);
 
-  buf->insert(buf->end(), chain_code_.begin(), chain_code_.end());
-  DCHECK_EQ(key->size(), 33u);
-  buf->insert(buf->end(), key->begin(), key->end());
+  buf.insert(buf.end(), chain_code_.begin(), chain_code_.end());
+  if (key.size() == 32) {
+    // 32-bytes private key is padded with a zero byte.
+    buf.insert(buf.end(), 0);
+  } else {
+    DCHECK_EQ(key.size(), 33u);
+  }
+  buf.insert(buf.end(), key.begin(), key.end());
 
-  DCHECK(buf->size() == SERIALIZATION_LEN);
-  return EncodeBase58Check(*buf);
+  DCHECK(buf.size() == SERIALIZATION_LEN);
+  return EncodeBase58Check(buf);
 }
```

Comment:
SecureZeroVectorDeleter actually might not work as expected. When bytes are inserted into wrapped vector it may reallocate leaving sensitive data as it is in memory. However that did work for Serialize as private keys always inserted last.
Still looks like a fragile construction.

Reply:
Since we don't use `SecureZeroVectorDeleter` anymore, can we remove it from `mem_utils`?

Reply:
Sure, nice catch